### PR TITLE
Move access control tests to each respective action tests #6356

### LINKED ISSUE #6356: Move access control tests to each respective action tests
It was that case long before and combined into one file in 65e9825 for reasons I could not find. I will give two reasons why this is not a wise move:
- When a new action is added, the dev needs to add new `*ActionTest` for it _and_ a new test segment in `AllAccessControlActionTest`. The latter are easier to forget, and indeed there are a few (luckily not a lot) missing access control tests (perhaps missing access control _check_ in production too?)
- `uri` has to be determined _for each page tested_. If the access control test is attached to its own `*ActionTest`, no such thing is needed. This might well be the very reason why `uri` needs to be static (a very dangerous design, luckily not in production).

What should then be done? `BaseActionTest` should admit protected abstract APIs `testExecuteAndPostProcess()` and `testAccessControl()`. This way, any new action tests added are forced to implement those two methods.

### PR DESCRIPTION
**Pull request by sujeet14108**
Fixes #6356

### CODE REVIEW

**Comment by sujeet14108** [comment]
Till now i have just made the `testAccessControl()` abstract and implement empty method in action classes which extends it. Do i also need to implement this method as per each class?

**Comment by leeyimin** [comment]
@sujeet14108 I believe so. You would need to at least move existing tests methods from AllActionsAccessControlTest to the correct action test and remove AllActionsAccessControlTest. You may want to open separate issues/PRs to implement the methods for action tests with no existing access control test.

**Comment by sujeet14108** [comment]
@leeyimin Can you help in solving failed test :(

**Comment by leeyimin** [comment]
@sujeet14108 For the failing admin action tests, you may want to remove the overloaded `prepareTestData` method if they are empty. For StudentCourseJoinAuthenticatedActionTest, you would need to use `removeAndRestoreDataBundle` to load the data in the datastore.

**Comment by sujeet14108** [comment]
@leeyimin I was in trouble with ` StudentCourseJoinAuthenticatedActionTest` mainly. Had applied what you were suggesting but still unable to solve it

File: src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java [hunk 1]
```diff
@@ -279,4 +282,50 @@ public void testExecuteAndPostProcess() throws Exception {
...
+
+        unregStudent1.googleId = "";
+        StudentsLogic.inst().updateStudentCascade(unregStudent1.email, unregStudent1);
+        verifyAccessibleForInstructorsOfOtherCourses(submissionParams);
+
+        unregStudent1.googleId = "";
+        StudentsLogic.inst().updateStudentCascade(unregStudent1.email, unregStudent1);
+        verifyAccessibleForAdminToMasqueradeAsInstructor(submissionParams);
+
+        ______TS("testStudentCourseJoinLegacyLink");
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java]: I believe this test case should be under StudentCourseJoinActionTest.

File: src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java [hunk 2]
```diff
@@ -279,4 +282,50 @@ public void testExecuteAndPostProcess() throws Exception {
     protected StudentCourseJoinAuthenticatedAction getAction(String... params) {
         return (StudentCourseJoinAuthenticatedAction) gaeSimulation.getActionObject(getActionUri(), params);
     }
+
+    @Override
+    @Test
+    protected void testAccessControl() throws Exception {
+        dataBundle = getTypicalDataBundle();
+        removeAndRestoreDataBundle(dataBundle);
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java]: Could make some changes to `StudentCourseJoinAuthenticatedTest.json` to include an unregistered user, rather than loading another data bundle.

File: src/test/java/teammates/test/cases/action/StudentFeedbackResultsPageActionTest.java [hunk 3]
```diff
@@ -22,6 +25,28 @@
...
+    public void classSetup() throws Exception {
+        addUnregStudentToCourse1();
+    }
+
+    @AfterClass
+    public void classTearDown() {
+        StudentsLogic.inst().deleteStudentCascade("idOfTypicalCourse1", "student6InCourse1@gmail.tmt");
+    }
+
+    private static void addUnregStudentToCourse1() throws Exception {
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentFeedbackResultsPageActionTest.java]: Can move method to BaseActionTest since it is reused a few times in several tests.

**Comment by sujeet14108** [comment]
@leeyimin  Can you help me in failing test of `StudentCourseJoinActionTest`

**Comment by leeyimin** [comment]
For `StudentCourseJoinActionTest`, you need to add the unreg student. To test the legacy link, you might want to create a new test class for the `testStudentCourseJoinLegacyLink` case since you need to return `Const.ActionURIs.STUDENT_COURSE_JOIN` when `getActionUri` is called.

File: src/test/java/teammates/test/cases/action/StudentCourseJoinLegacyLinkTest.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+import teammates.common.datatransfer.attributes.StudentAttributes;
+import teammates.common.util.Const;
+import teammates.common.util.StringHelper;
+import teammates.logic.core.StudentsLogic;
+import teammates.ui.controller.Action;
+import teammates.ui.controller.StudentCourseJoinAction;
+
+/**
+ * SUT: {@link StudentCourseJoinAction}.
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentCourseJoinLegacyLinkTest.java]: You may want to add more information about the purpose of this test since it does not explain why StudentCourseJoinActionTest is insufficient.

File: src/test/java/teammates/test/cases/action/StudentCourseJoinLegacyLinkTest.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+ */
+public class StudentCourseJoinLegacyLinkTest extends BaseActionTest {
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentCourseJoinLegacyLinkTest.java]: Maybe rename it as `StudentCourseJoinActionLegacyTest` so that it is clearer from the name which action it is testing.

File: src/test/java/teammates/test/cases/action/StudentCourseJoinLegacyLinkTest.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Override
+    protected Action getAction(String... params) {
+        return (StudentCourseJoinAction) gaeSimulation.getActionObject(getActionUri(), params);
+    }
+
+    @Override
+    protected void testExecuteAndPostProcess() throws Exception {
+        // TODO Auto-generated method stub
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentCourseJoinLegacyLinkTest.java]: Remove the todo and replace with a comment that this is not required.

File: src/test/java/teammates/test/cases/action/StudentCourseJoinLegacyLinkTest.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        StudentsLogic.inst().deleteStudentCascade("idOfTypicalCourse1", "student6InCourse1@gmail.tmt");
+    }
+
+    @Override
+    protected String getActionUri() {
+        return Const.ActionURIs.STUDENT_COURSE_JOIN;
+    }
+
+    @Override
+    protected Action getAction(String... params) {
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentCourseJoinLegacyLinkTest.java]: Change return type to `StudentCourseJoinAction`.

File: src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java [hunk 8]
```diff
@@ -26,14 +27,16 @@ protected String getActionUri() {
...
         // extra test data used on top of typical data bundle
-        removeAndRestoreDataBundle(dataBundle);
     }
 
     @Override
     @Test
     public void testExecuteAndPostProcess() throws Exception {
+        dataBundle = loadDataBundle("/StudentCourseJoinAuthenticatedTest.json");
+        removeAndRestoreDataBundle(dataBundle);
+
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java]: Revert the changes here if it is no longer required.

File: src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java [hunk 9]
```diff
@@ -26,14 +27,16 @@ protected String getActionUri() {
     }
 
     @BeforeClass
-    public void classSetup() {
+    public void classSetup() throws Exception {
         // extra test data used on top of typical data bundle
-        removeAndRestoreDataBundle(dataBundle);
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java]: Revert the changes here if it is no longer required.

File: src/test/java/teammates/test/cases/action/StudentFeedbackResultsPageActionTest.java [hunk 10]
```diff
@@ -213,4 +226,23 @@ public void testExecuteAndPostProcess() throws Exception {
...
+
+        String[] submissionParams = new String[] {
+                Const.ParamsNames.COURSE_ID, session1InCourse1.getCourseId(),
+                Const.ParamsNames.FEEDBACK_SESSION_NAME,
+                session1InCourse1.getFeedbackSessionName()
+        };
+
+        verifyOnlyStudentsOfTheSameCourseCanAccess(submissionParams);
+
+        // TODO: test no questions -> redirect after moving detection logic to
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentFeedbackResultsPageActionTest.java]: You may want to investigate this TODO and see if this is still relevant and remove it if it is not.

File: src/test/resources/data/StudentCourseJoinAuthenticatedTest.json [hunk 11]
```diff
@@ -134,6 +134,15 @@
...
     },
+    "student1InUnregisteredCourse": {
+      "email": "student1InUnregisteredCourse@gmail.tmt",
+      "course": "idOfUnregisteredCourse",
+      "name": "student1 In unregisteredCourse",
+      "comments": "",
+      "team": "Team 1",
+      "section": "Section 1",
+      "key": "regKeyForStuNotYetJoinCourse"
+    },
```
Comment (leeyimin) [inline src/test/resources/data/StudentCourseJoinAuthenticatedTest.json]: I believe you do not have to add this additional student actually. You can refer to the desired student by using `getTypicalDataBundle()`.

File: src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java [hunk 12]
```diff
@@ -279,4 +282,34 @@ public void testExecuteAndPostProcess() throws Exception {
     protected StudentCourseJoinAuthenticatedAction getAction(String... params) {
         return (StudentCourseJoinAuthenticatedAction) gaeSimulation.getActionObject(getActionUri(), params);
     }
+
+    @Override
+    @Test
+    protected void testAccessControl() throws Exception {
+
+        StudentAttributes unregStudent1 = dataBundle.students.get("student1InUnregisteredCourse");
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java]: You can use `getTypicalDataBundle()` to get the desired student in that data set instead of adding it to `StudentCourseJoinAuthenticatedTest.json`.

File: src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java (hunk 8 above)
Comment (sujeet14108) [inline src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java]: load of dataBundle is required here as it is executed after access control and in accessControl we have changes dataBundle to `typicaldataBundle`.
May we can remove Line 22 and add data bundle here

File: src/test/java/teammates/test/cases/action/StudentFeedbackResultsPageActionTest.java (hunk 10 above)
Comment (sujeet14108) [inline src/test/java/teammates/test/cases/action/StudentFeedbackResultsPageActionTest.java]: @leeyimin I have don't know what is the comment suggesting so keeping it still till the we find out what it means

**Comment by sujeet14108** [comment]
Ready for review :)

File: src/test/java/teammates/test/cases/action/StudentCourseJoinActionLegacyTest.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+import org.testng.annotations.Test;
+
+import teammates.common.datatransfer.attributes.StudentAttributes;
+import teammates.common.util.Const;
+import teammates.common.util.StringHelper;
+import teammates.logic.core.StudentsLogic;
+import teammates.ui.controller.StudentCourseJoinAction;
+
+/**
+ * Test Course Join Link for Unregistered Student
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentCourseJoinActionLegacyTest.java]: How about "To test legacy course join link for unregistered students."? The other words are unneccessarily capitalised.

File: src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java [hunk 14]
```diff
@@ -26,14 +27,15 @@ protected String getActionUri() {
...
-    public void classSetup() {
+    public void classSetup() throws Exception {
         // extra test data used on top of typical data bundle
         removeAndRestoreDataBundle(dataBundle);
     }
 
     @Override
     @Test
     public void testExecuteAndPostProcess() throws Exception {
+        dataBundle = loadDataBundle("/StudentCourseJoinAuthenticatedTest.json");
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java]: Is this still necessary?

File: src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java [hunk 15]
```diff
@@ -26,14 +27,15 @@ protected String getActionUri() {
     }
 
     @BeforeClass
-    public void classSetup() {
+    public void classSetup() throws Exception {
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java]: Is this change necessary?

Comment (sujeet14108) [inline src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java]: No need of it now. Reverting

File: src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java (hunk 14 above)
Comment (sujeet14108) [inline src/test/java/teammates/test/cases/action/StudentCourseJoinAuthenticatedActionTest.java]: Load of dataBundle is required here as it is executed after access control and in accessControl we have changed dataBundle to typicaldataBundle

**Comment by sujeet14108** [comment]
@leeyimin  Please check

**Comment by sujeet14108** [comment]
Is there any problem in merge :sweat_smile: ?
